lektor/watcher.py

```diff
@@ -2,59 +2,86 @@
 import queue
 import time
 
+import click
 from watchdog.events import DirModifiedEvent
 from watchdog.events import FileMovedEvent
 from watchdog.events import FileSystemEventHandler
 from watchdog.observers import Observer
+from watchdog.observers.polling import PollingObserver
 
 from lektor.utils import get_cache_dir
 
-# Alias this as this can be called during interpreter shutdown
-_Empty = queue.Empty
-
 
 class EventHandler(FileSystemEventHandler):
-    def __init__(self, callback=None):
-        if callback is not None:
-            self.queue = None
-            self.callback = callback
-        else:
-            self.queue = queue.Queue()
-            self.callback = self.queue.put
+    def __init__(self):
+        self.queue = queue.Queue()
 
     def on_any_event(self, event):
         if not isinstance(event, DirModifiedEvent):
-            path = (
-                event.dest_path if isinstance(event, FileMovedEvent) else event.src_path
-            )
-            item = (time.time(), event.event_type, path)
-            if self.queue is not None:
-                self.queue.put(item)
+            if isinstance(event, FileMovedEvent):
+                path = event.dest_path
             else:
-                self.callback(*item)
+                path = event.src_path
+            self.queue.put((time.time(), event.event_type, path))
 
 
 class BasicWatcher:
-    def __init__(self, paths, callback=None):
-        self.event_handler = EventHandler(callback=callback)
-        self.observer = Observer()
-        for path in paths:
-            self.observer.schedule(self.event_handler, path, recursive=True)
-        self.observer.setDaemon(True)
+    def __init__(self, paths, observer_classes=(Observer, PollingObserver)):
+        self.event_handler = EventHandler()
+        self.paths = paths
+        self.observer_classes = observer_classes
+        self.observer = None
+
+    def start(self):
+        untried = set(self.observer_classes)
```

Comment:
This set seems to be mainly used to check whether there's still an observer left in the except clause or whether we need to re-raise. The fact that the set removes duplicates also has some weird effects in that a failing observer class would be tried multiple times if given multiple times at the start or middle of the list of observer classes, but it would only be tried once if given multiple times at the end of that list.

Maybe with an iterator, this could be handled in the same loop (two other nice side effects: fails loudly for empty observer_classes and gives us access to the next class in the except handler). 

```suggestion
        observers = iter(self.observer_classes)
        observer_class = next(observers)
        while observer_class: [...]
```

and then a `observer_class = next(observers, None)` in the except.

Reply:
The real (though esoteric) case that this was concerned with was to prevent issuing a spurious "falling back to polling" message when creating the default `Observer` fails on systems where `Observer` is `PollingObserver`.

I've refactored the start-observer loop.  See if you like.